doc/development/guide/documentation.rst

```diff
@@ -1,39 +1,455 @@
+
+        **Example**
+
+        Minimal example with 1 or 2 code blocks (required).
+
+        .. UsageDetails::
+
+            More complicated use cases, options, and larger code blocks (optional).
+
+        **Related tutorials**
+
+        Links to any relevant PennyLane demos (optional).
+        """
+
+
+For predominantly developer-facing functions and methods (e.g., private functions and methods),
+the ``Example`` section is not required. In addition, if the function is very short and its use
+is apparent from the signature, it is sufficient for the docstring to simply consist of the single
+line summary sentence.
+
+Some notes on the above structure:
+
+* **Function summary:** The summary should provide a brief, basic description of the function. Do not
+  include the function or argument names in the summary. It is important that the initial summary
+  be a single sentence; Sphinx will truncate the summary at the first period otherwise.
+
+* **Arguments:** To describe named function arguments in the signature. If the type is a PennyLane class or
+  function, use the syntax ``(~.Operation)``; do **not** use a Sphinx role such as ``:class:`` or ``:func:``.
+  The level of type description is up to the discretion of the code author and reviewers; acceptable
+  examples include ``(array)``, ``(array[float])``, ``(dict)``, ``(dict[str, int])``, ``(Sequence[int] or int)``.
+  Note that square brakets are used to define the required types of container elements.
+
+  *Note:* if the argument has a default value, *do not* include it in the description if it is declared
+  in the signature. Sphinx will automatically extract the default and render it with the argument
+  description.
+
+* **Keyword Arguments:** To describe variable length named keyword arguments provided via
+  ``**kwargs``. Provide the default values if relevant.
+
+* **Returns:/Yields:** To describe the return (yielded) value of the function (generator). This may
+  be omitted if the function signature/summary already sufficiently describe both the return value
+  and type.
+
+* **See also:** Provide optional links to relevant pages, functions, classes, methods, etc.
+
+* **Example:** To provide a minimal working example showing basic usage of the function. The example
+  should be *minimal* (reduce line counts where possible), but complete; a reader should be able to
+  copy and paste the example and get the same output. See :ref:`code_examples` for guidelines writing
+  useful code examples in docstrings.
+
+* **Usage details:** To provide a more complicated usage details showing different edge cases and
+  advanced usage of the function, as well as implementation details. This section of the docstring
+  is *collapsed by default* in the rendered Sphinx documentation to avoid overwhelming the reader with
+  information, however will be displayed in full to users in the terminal/Jupyter notebooks, so
+  do not let this section become too long.
+
```

Comment:
I think this is very Numba specific for now. Most decorators, especially the ones we code ourselves, use `functools.wraps` so that the wrapped function docstring is copied into the decorated function. Unfortunately, Numba doesn't seem to do this.

Regardless, since this is very Sphinx/Numba dependent, and could be fixed with updates, I don't think it fits in the general guidelines document.

Reply:
An exception would be if we converge on a temporary solution for Numba decorated functions. If it differs from the above procedure, then we should add it to this page

Reply:
I see. I agree that it might not be needed in that case. I guess it depend on how often developers might use decorators that are not using `functool.wraps`.